### DISCUSSION TITLE
Installation as a C++ novice

### DISCUSSION THREAD

**Discussion by edvinsmineikis** [alandefreitas/matplotplusplus#134] [Q&A]
[error.txt](https://github.com/alandefreitas/matplotplusplus/files/6292139/error.txt)

Greetings!

Before I say anything, do realize I have no experience with C++ projects. As far as I can tell, matplotplusplus offers a collection of files you can call with
`#include <matplot/matplot.h`
in any other file you want to compile.

Considering the guide on README.md, and the fact I have never used CMake, after about an hour of trying to get the library working I have failed.

Being a Linux user and C++ noob, I put the whole "matplot" directory into my _/usr/include_ directory, hoping that upon including it into any other C++ code it would work just fine. My logic is that there is no point in pre-compiling any files if they will be the includes of another file that will be compiled. As far as I have researched, CMake is a tool to make life easier for those who want to compile a project consisting of many C or C++ files, so I guessed I dont have to do it.

Upon compiling my code with g++ :
```
#include <matplot/matplot.h>
#include <iostream>

using namespace std; // Ive also tried "matplot" and no namespace

void main() {
        std::cout<"Test";
}
```
I get a lot of similar errors that a function is not a member of 'std'.
`/usr/include/matplot/util/concepts.h:66:40: error: ‘is_same_v’ is not a member of ‘std’; did you mean ‘is_same’?`
All other errors probably come as side-symptoms to this problem. I suppose I do not understand namespaces well enough.

Any type of advice is greatly appreciated.

**Comment by lacc97** [comment]
Those errors happen because you are compiling as C++14 or lower, and matplot++ requires C++17. This is the sort of thing CMake takes care of (or at least explicitly tells you if you are missing this).

In any case, your nonstandard installation is not going to work because matplot++ is not a header-only library. Even if you somehow manage to get things to compile you will have link errors, and fixing those means you might as well rewrite the entire build system.

I think it would be much more useful for everyone if you explained what you have tried with setting up CMake and what failed.

**Reply by edvinsmineikis** [reply]
Your answer helped. I have never built a project with CMake before so it was something new, but in the end I made it work. Thank you!

**Comment by alandefreitas** [comment]
Hi @Zengit123 , thank you for your feedback!

I think the problem here is with the assumption that 

> matplotplusplus offers a collection of files you can call with `#include <matplot/matplot.h` in any other file you want to compile.

That would be the case if matplotplusplus were a header-only library, but it's not. 
Not all libraries are good candidates as header-only libraries.

Matplotplusplus is a library you need to compile, so this won't work either

> I put the whole "matplot" directory into my /usr/include directory, hoping that upon including it into any other C++ code it would work just fine. 

I understand how the way C++ works might be confusing especially when you come from another language. 

For instance, when you say

> My logic is that there is no point in pre-compiling any files if they will be the includes of another file that will be compiled. 

it seems like you are assuming the files are only being included or amalgamated as if they were header-only, and you could bypass the building system. It seems like the assumption here is that all the building system does is amalgamate and include these files.

The building system is actually doing a lot more than that under the hood. The build system would identify your system, your compiler, download or identify the appropriate dependencies, set compile-time definitions, ...  So doing all of that without the help of CMake is not very simple at all.

One possibility you can consider is a package manager, but the problem here is that most libraries just won't be there. If I'm not wrong, Conan is the largest package manager but it has less than a thousand packages, while C++ has been around for more than 40 years. 
Due to historical and performance reasons, [most C++ programmers just manually manage](https://www.cin.ufpe.br/~jhcp/publica/sac2018cpp-pre.pdf) their dependencies.

I'm still thinking of an easier solution for this problem, especially considering programmers coming from other languages, because my main goal here to make C++ simpler for scientific computing and a lot of people are coming from python or even matlab. This solution would be something like npm/pip for C++. I've already done some of the work for a pre-release version, but that's not something so simple to achieve in C++, because you can't just amalgamate the source files (for a number of reasons, including the influence of different build systems in the process), and this tool would need to actually resolve the dependency conflicts (unlike npm, that duplicates conflicting dependencies). 

For now, the only thing I can recommend is to not bypass the "integration" section in the documentation. It's unfortunately not as simple as npm/pip/... but it's much more convenient than bypassing the build system, especially if you don't have a lot of experience with that yet.

**Reply by edvinsmineikis** [reply]
Thank you for the detailed answer. Have done a bit of research on how compiling and linking works, and why we shouldnt just include plain cpp files, now I understand better :)
I can also say that the Cmake guidelines in the readme caused a bit of confusion because I had no idea what "my project" was supposed to mean. Because I didnt know I had to make a new CMakeLists.txt file to include the library linkage, I had to go through a few tutorials to understand where to put the required cmake commands. Maybe some C++ newcomers also had this issue.

**Reply by alandefreitas** [reply]
Yes. I understand how all of that can be confusing. Some of it is what allows C++ to be efficient. Some of it is due to historical reasons. It's much easier to come up with constraints on these complexities when coming up with a new language. 

What programming language are you coming from? The CMake reference is pretty good but it's not very didactic. You might use [this](https://lrita.github.io/images/posts/cplusplus/mastering-cmake.pdf) as a guide but it might be too much information for what you need. As visual studio also adopts CMake, it seems like it's [becoming](https://www.jetbrains.com/lp/devecosystem-2019/cpp/) the dominant build system in the C++ world but that's far from unanimous.

Most IDEs will give you a starting CMakeLists.txt file when you create a new project. So another option is to use this template as a starting point, understand how the compiling and linking processes work, and then check the CMake reference when you need this or that feature.

**Reply by edvinsmineikis** [reply]
I come more from hardware side and python, electrical engineer basically. I can make simple automation stuff with linux+python including webscraping, serial communications, networking and general system things, but never actually wrote any complex applications. Knowing C++ is one of most valuable skills in IT, in my opinion, decided to step into open-source to catch up to speed.
Also I use the Terminal to do everything instead of IDE's. Found that CMake is very handy because I just had to add a few lines to a CMakeList.txt, do "cmake" and then do "make". Every time I change anything, I just do "make" again and it rebuilds the project efficiently, very handy tools for applying changes.

**Reply by alandefreitas** [reply]
It's good to know that kind of thing so I can improve things over here. I also come from electrical engineering and still use python eventually. 

Setting up a C/C++ project might seem unnecessarily painful at first, but it's worth it once you can achieve the things directly in C/C++ rather than using Python which is probably just wrapping stuff in C/C++ anyway. 

We'll solve these unnecessary problems eventually, but it's still worth it as it is.
